File: python/sdk/src/windy_drops/lib/scaffold.py
"""scaffold.py — copy a per-type starter directory from
examples/<type>-minimal/ into a user-specified path, rewriting the placeholder
`id` to match the target dir name.

The examples directory is the source of truth — updating an example
automatically updates what `windy-drops new` produces. Mirrors the TS SDK's
scaffold.ts so both languages produce byte-identical scaffolds.
"""

from __future__ import annotations

import os
import re
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Final

RESERVED_DROP_TYPES: Final[tuple[str, ...]] = (
    "control-panel-template",
    "skill",
    "tool",
    "theme",
    "voice-pack",
    "workflow",
)


@dataclass(frozen=True)
class ScaffoldResult:
    examples_dir: Path
    target_dir: Path
    files_copied: list[str]


def _find_examples_root() -> Path:
    """Locate the repo's examples/ directory by walking up from this file.

    When installed via pip, the examples ship inside the package bundle (a
    sibling `examples/` directory). When running from source, the examples
    live a few levels up at <repo_root>/examples/.
    """
    here = Path(__file__).resolve().parent
    candidates = [
        here.parent.parent.parent.parent.parent / "examples",  # source tree
        here.parent.parent.parent.parent / "examples",          # python/ root
        here.parent.parent.parent / "examples",                 # src/ root
        here.parent / "examples",                               # bundled
    ]
    for c in candidates:
        if c.exists() and c.is_dir():
            return c
    raise FileNotFoundError(
        f"Could not locate examples/ directory. Searched: "
        + ", ".join(str(c) for c in candidates)
    )


def scaffold(*, type: str, path: str | os.PathLike[str]) -> ScaffoldResult:
    """Scaffold a starter drop of the given type at the given path."""
    if type not in RESERVED_DROP_TYPES:
        raise ValueError(
            f"Unknown drop type: {type}. Must be one of: "
            + ", ".join(RESERVED_DROP_TYPES)
        )

    examples_root = _find_examples_root()
    examples_dir = examples_root / f"{type}-minimal"
    if not examples_dir.exists():
        raise FileNotFoundError(
            f"No starter scaffold for type {type} at {examples_dir}"
        )

    target_dir = Path(path).resolve()
    if target_dir.exists():
        if any(target_dir.iterdir()):
            raise FileExistsError(f"Target directory not empty: {target_dir}")
    else:
        target_dir.mkdir(parents=True)

    # Recursively copy.
    for entry in examples_dir.iterdir():
        if entry.is_dir():
            shutil.copytree(entry, target_dir / entry.name)
        else:
            shutil.copy2(entry, target_dir / entry.name)

    # Rewrite the placeholder `id` to a target-dir-named id.
    skill_path = target_dir / "SKILL.md"
    if skill_path.exists():
        body = skill_path.read_text(encoding="utf-8")
        slug = target_dir.name or "my-drop"
        rewritten = re.sub(
            r"^id: your-handle-(.+)$",
            f"id: your-handle-{slug}",
            body,
            flags=re.MULTILINE,
        )
        skill_path.write_text(rewritten, encoding="utf-8")

    files_copied = sorted(
        str(p.relative_to(target_dir))
        for p in target_dir.rglob("*")
        if p.is_file()
    )

    return ScaffoldResult(
        examples_dir=examples_dir,
        target_dir=target_dir,
        files_copied=files_copied,
    )

File: packages/sdk/src/lib/scaffold.ts
// scaffold.ts — copy a per-type starter directory from examples/<type>-minimal/
// into a user-specified path, rewriting placeholders.
//
// The examples directory is the source of truth — updating an example
// automatically updates what `windy-drops new` produces.

import { cp, mkdir, readdir, readFile, writeFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

export const RESERVED_DROP_TYPES = [
  "control-panel-template",
  "skill",
  "tool",
  "theme",
  "voice-pack",
  "workflow",
] as const;

export type DropType = (typeof RESERVED_DROP_TYPES)[number];

export interface ScaffoldOptions {
  type: string;
  path: string;
}

export interface ScaffoldResult {
  examplesDir: string;
  targetDir: string;
  filesCopied: string[];
}

// Locate the repo's examples/ directory. Walks up from this file looking
// for `examples/<type>-minimal/`. When the SDK is installed via npm, the
// examples ship inside the package bundle (`packages/sdk/examples/` mirrored
// at publish time). For now we resolve relative to the source tree.
function findExamplesRoot(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  // Try: <here>/../../examples (when running from packages/sdk/dist/lib/)
  // and: <here>/../examples (when bundled as a single-folder package)
  const candidates = [
    resolve(here, "../../../../examples"),
    resolve(here, "../../../examples"),
    resolve(here, "../../examples"),
    resolve(here, "../examples"),
  ];
  for (const c of candidates) {
    if (existsSync(c)) return c;
  }
  throw new Error(
    `Could not locate examples/ directory. Searched: ${candidates.join(", ")}`,
  );
}

export async function scaffold(opts: ScaffoldOptions): Promise<ScaffoldResult> {
  if (!RESERVED_DROP_TYPES.includes(opts.type as DropType)) {
    throw new Error(
      `Unknown drop type: ${opts.type}. Must be one of: ${RESERVED_DROP_TYPES.join(", ")}`,
    );
  }

  const examplesRoot = findExamplesRoot();
  const examplesDir = resolve(examplesRoot, `${opts.type}-minimal`);

  if (!existsSync(examplesDir)) {
    throw new Error(`No starter scaffold for type ${opts.type} at ${examplesDir}`);
  }

  const targetDir = resolve(opts.path);
  if (existsSync(targetDir)) {
    const contents = await readdir(targetDir).catch(() => []);
    if (contents.length > 0) {
      throw new Error(`Target directory not empty: ${targetDir}`);
    }
  } else {
    await mkdir(targetDir, { recursive: true });
  }

  await cp(examplesDir, targetDir, { recursive: true });

  // Rewrite the placeholder `id` to a target-dir-named id so the scaffold
  // isn't a duplicate-id-on-publish footgun.
  const skillPath = resolve(targetDir, "SKILL.md");
  const skillBody = await readFile(skillPath, "utf-8");
  const slug = targetDir.split("/").pop() ?? "my-drop";
  const rewritten = skillBody.replace(
    /^id: your-handle-(.+)$/m,
    `id: your-handle-${slug}`,
  );
  await writeFile(skillPath, rewritten, "utf-8");

  const filesCopied = await listFiles(targetDir);
  return { examplesDir, targetDir, filesCopied };
}

async function listFiles(dir: string): Promise<string[]> {
  const entries = await readdir(dir, { withFileTypes: true, recursive: true });
  return entries
    .filter((e) => e.isFile())
    .map((e) => `${e.parentPath ?? e.path}/${e.name}`)
    .sort();
}
